### ISSUE TITLE
fatal error: xmmintrin.h: No such file or directory

### ISSUE THREAD

**Issue by RoryZahedi** [esonghori/TinyGarble#25]
Hi, I cloned tiny garble onto my pynq FPGA, and when I ran make I ran into the following issue.

![image](https://user-images.githubusercontent.com/19734560/219980267-2eed48c5-8051-4751-80f5-658014777073.png)
It seems the issue may be because of difference in architecture, but I was wondering if there are any workarounds or will support be added for ARM architectures?

**Comment by esonghori** [comment]
You are right about the cause: TinyGarble's garbling core is x86-specific. Wire labels are `__m128i` blocks (`crypto/block.h`) and the fixed-key block cipher uses AES-NI intrinsics (`crypto/aes.h`), and the build adds `-march=native`. There is no ARM path today, so `xmmintrin.h` is simply not there on your PYNQ board.

I have documented the requirement rather than papered over it — `README.md` now states up front that an x86-64 CPU with AES-NI is needed, so this fails at the dependency list instead of deep inside a build error (`main`, a6191f1).

A real ARM port means replacing the SSE/AES-NI layer in `crypto/block.h` and `crypto/aes.h` with NEON and ARM crypto extension equivalents. That is a contained piece of work — those two headers are where essentially all of the architecture dependence lives — but it is more than a workaround and it is not done. I am leaving this issue open to track it. In the meantime, the only option is to run TinyGarble on an x86-64 host.

**Comment by esonghori** [comment]
This is now implemented. TinyGarble builds and runs on AArch64 as of df963af on `main`.

Your diagnosis was right: the garbling core is written against x86 SSE and AES-NI, so `xmmintrin.h` simply does not exist on your board. Rather than fork the protocol into an x86 version and an ARM version, `crypto/neon_compat.h` supplies the ~30 intrinsics this codebase uses on top of NEON and the ARMv8 crypto extensions, and `crypto/block.h` and `crypto/aes.h` include it instead of the x86 headers on AArch64. Nothing in `garbled_circuit/` changed.

Three details, in case they are useful to anyone doing similar work:

- ARM and x86 slice an AES round differently. `AESE` does AddRoundKey, SubBytes and ShiftRows with MixColumns split out into `AESMC`, whereas x86's `AESENC` does the whole round including the key. Passing a zero key to `AESE` and XORing the round key afterwards reproduces x86 exactly.
- ARMv8 has no counterpart to `AESKEYGENASSIST`, so the key schedule helper is done in software against the FIPS-197 S-box.
- Two x86 dependencies lived outside the crypto headers: the `RDTSC` macro in `util/common.h`, which now reads `cntvct_el0` on AArch64, and `a23/sha3/byte_order.h`, which knew about `__arm__` but not `__aarch64__` and so failed with "Can't detect CPU architechture".

**On correctness.** Bit-exactness matters more here than it might appear: an x86 garbler and an ARM evaluator have to agree on every wire label, so a shim that is merely self-consistent would pass every single-machine test and then fail the moment two different machines talk to each other. So:

- AES-128 matches the FIPS-197 C.1 and SP800-38A F.1.1 known-answer vectors, key schedule and decrypt path included.
- A program exercising every shimmed primitive — block operations, the SSE-based PRNG, all 11 AES round keys, ECB blocks — produces byte-identical output on x86-64 and AArch64. 34 of 34 lines match.
- All 8 `ctest` suites pass natively on AArch64.
- An actual cross-architecture protocol run works: Alice on Apple silicon against Bob on x86-64 over TCP returns the correct answer for `sum_8bit_1cc`, for `hamming_32bit_32cc` both with and without `--low_mem_foot` (the two garbling implementations are separate code paths), and for `mult_8bit_8cc`.

The x86 build is unchanged and its tests still pass.

**One caveat for your board specifically.** This is verified on AArch64 (ARM64). A PYNQ's ARM core may be 32-bit ARMv7 (`__arm__`), depending on the board — the Z1 and Z2 are Zynq-7000, which is 32-bit Cortex-A9 with NEON but *no* crypto extensions. In that case this will not help you directly: the shim is guarded to AArch64, and without hardware AES the fixed-key blockcipher would need a software AES, which would be far slower. If you are on a Zynq UltraScale+ (64-bit, ARMv8) it should work. Please reopen and say which board you have if you hit that, and note the AES issue.
